calico.go

```diff
@@ -44,80 +45,111 @@ func init() {
-	}
-	if conf.Nodename != "" {
-		nodename = conf.Nodename
-	}
 }
 
 func cmdAdd(args *skel.CmdArgs) error {
 	// Unmarshal the network config, and perform validation
-	conf := NetConf{}
+	conf := types.NetConf{}
 	if err := json.Unmarshal(args.StdinData, &conf); err != nil {
 		return fmt.Errorf("failed to load netconf: %v", err)
 	}
 
-	cniVersion := conf.CNIVersion
+	utils.ConfigureLogging(conf.LogLevel)
 
-	ConfigureLogging(conf.LogLevel)
+	// Allow the nodename to be overridden by the network config
+	nodename = utils.DetermineNodename(conf)
 
-	workload, orchestrator, err := GetIdentifiers(args)
+	// Extract WEP identifiers such as pod name, pod namespace (for k8s), containerID, workload name.
+	epIDs, err := utils.GetIdentifiers(args, nodename)
 	if err != nil {
 		return err
 	}
 
-	logger := CreateContextLogger(workload)
+	logrus.WithField("EndpointIDs", epIDs).Info("Extracted identifiers")
+	logrus.WithField("NetConfg", conf).Info("Loaded CNI NetConf")
 
-	// Allow the nodename to be overridden by the network config
-	updateNodename(conf, logger)
+	calicoClient, err := utils.CreateClient(conf)
+	if err != nil {
+		return err
+	}
 
-	logger.WithFields(log.Fields{
-		"Orchestrator": orchestrator,
-		"Node":         nodename,
-		"Workload":     workload,
-		"ContainerID":  args.ContainerID,
-	}).Info("Extracted identifiers")
+	// Remove the endpoint from the epIDs so we can get a WEP name prefix.
+	epIDs.Endpoint = ""
 
-	logger.WithFields(log.Fields{"NetConfg": conf}).Info("Loaded CNI NetConf")
-	calicoClient, err := CreateClient(conf)
+	// Calculate the workload name from the WEP specific identifiers
+	// for the given orchestrator.
+	wepPrefix, err := epIDs.CalculateWorkloadEndpointName(true)
 	if err != nil {
-		return err
+		return fmt.Errorf("error constructing WorkloadEndpoint prefix: %s", err)
 	}
 
-	// Always check if there's an existing endpoint.
-	endpoints, err := calicoClient.WorkloadEndpoints().List(api.WorkloadEndpointMetadata{
-		Node:         nodename,
-		Orchestrator: orchestrator,
-		Workload:     workload})
+	ctx := context.Background()
+
+	// Check if there's an existing endpoint.
+	endpoints, err := calicoClient.WorkloadEndpoints().List(ctx, options.ListOptions{Name: wepPrefix, Namespace: epIDs.Namespace, Prefix: true})
 	if err != nil {
 		return err
 	}
 
-	logger.Debugf("Retrieved endpoints: %v", endpoints)
+	var logger *logrus.Entry
+	if epIDs.Orchestrator == "k8s" {
+		logger = logrus.WithFields(logrus.Fields{
+			"WorkloadEndpoint": fmt.Sprintf("%s%s", wepPrefix, epIDs.Endpoint),
+			"ContainerID":      epIDs.ContainerID,
+			"Pod":              epIDs.Pod,
+			"Namespace":        epIDs.Namespace,
+		})
+	} else {
+		logger = logrus.WithFields(logrus.Fields{
+			"ContainerID": epIDs.ContainerID,
+		})
+	}
+
+	logger.Debugf("Retrieved list of endpoints: %v", endpoints)
 
 	var endpoint *api.WorkloadEndpoint
-	if len(endpoints.Items) == 1 {
-		endpoint = &endpoints.Items[0]
-	}
 
```

Comment:
I think you need to put a detailed comment here explaining exactly what you are doing here.